### DISCUSSION TITLE
Improve block rendering

### DISCUSSION THREAD

**Discussion by amargoCactus** [area17/twill#2447] [Ideas]
## Summary

When many blocks that use medias, translations, repeaters and any feature that needs to query a model are render, it strongly affects the server response performance (TTFB) and debugbar indicates a huge number of queries, it affects the performance a lot, in my case the page is taking 13 seconds for TTFB

## Describe the solution you'd like

Maybe eager loading or caching?

## Additional context

![image](https://user-images.githubusercontent.com/36394543/204709123-e8ea4a88-53c7-46f9-941b-714065fbe5d9.png)

**Comment by ifox** [comment]
Hi @amargoCactus,

It seems like you're saying that the amount of queries is strictly due to Twill, so I have a couple questions for you.

-  Is that really the case? How are you getting the homepage model and are you doing any eager loading there on your end?
- How many blocks and browser attached items are you dealing with? 
- How much of those 13s is actually spent on db queries?

Regarding caching, I don't think it is Twill's responsibility to handle caching of the data you render on your frontend. Laravel has a lot of caching options though, both internal and external, and they all work very well with Twill, so I would suggest looking into that. We have websites that require large amount of queries for certain things that would result in unacceptable TTFB numbers, but with a CDN or response cache, we can still serve them at sub 50ms TTFB.

**Comment by amargoCactus** [comment]
> How are you getting the homepage model and are you doing any eager loading there on your end

i tried to eager load my `Page` model in the controller and it didn't improve, because the `A17\Twill\Helpers\BlockRenderer` class does the queries anyway. I load the pages like this:

```php
//routes/web.php
Route::group([
    'prefix' => LaravelLocalization::setLocale(),
    'middleware' => ['localize', 'localeSessionRedirect', 'localizationRedirect', 'localeViewPath'],
], function () {
    Route::get('{slug?}', [PageController::class, 'index'])->name('pages.index');
})
```

```php
<?php

namespace App\Http\Controllers\Site;

use App\Models\Page;
use App\Http\Controllers\Controller;
use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;

class PageController extends Controller
{
    public function defaultPage()
    {
        $default = Page::first();
        return redirect()->route('pages.index', $default->slug);
    }

    public function index($slug)
    {
        try {
            $data = Page::forSlug($slug)->first();

            if (!$data) {
                throw new NotFoundHttpException();
            }
        } catch (\Exception $e) {
            throw new NotFoundHttpException();
        }

        return view('site.pages', ['data' => $data]);
    }
}
```

```
//resources/site/pages.blade.php
@extends('site.layouts.site')

@section('content')
    {!! $data->renderBlocks() !!}
@endsection
```

> How many blocks and browser attached items are you dealing with

The homepage for the moment has 6 blocks

> How much of those 13s is actually spent on db queries

![image](https://user-images.githubusercontent.com/36394543/204854027-92413fa5-7f27-4f28-be1d-6a5e98c4fc0c.png)

> Regarding caching, I don't think it is Twill's responsibility to handle caching of the data you render on your frontend

Agree, but i think it would be nice if `BlockRenderer` class if the class may reduce the number of queries with eager loading
